src/ray/object_manager/plasma/store.cc

```diff
@@ -190,11 +192,22 @@ uint8_t* PlasmaStore::AllocateMemory(size_t size, bool evict_if_full, MEMFD_TYPE
     }
     // Tell the eviction policy how much space we need to create this object.
     std::vector<ObjectID> objects_to_evict;
-    bool success = eviction_policy_.RequireSpace(size, &objects_to_evict);
+    int64_t space_needed = eviction_policy_.RequireSpace(size, &objects_to_evict);
     EvictObjects(objects_to_evict);
-    // Return an error to the client if not enough space could be freed to
-    // create the object.
-    if (!success) {
+    // More space is still needed. Try to spill objects to external storage to
+    // make room.
+    if (space_needed > 0) {
+      if (spill_objects_callback_) {
+        // Object spilling is asynchronous so that we do not block the plasma
+        // store thread. Therefore the client must try again, even if enough
+        // space will be made after the spill is complete.
+        // TODO: Only respond to the client with OutOfMemory if we could not
+        // make enough space through spilling. If we could make enough space,
+        // respond to the plasma client once spilling is complete.
+        static_cast<void>(spill_objects_callback_(space_needed));
```

Comment:
I might misunderstand here, but doesn't this mean raylet can try creating new objects or pulling new objects "before" all objects are actually spilled, and it can lead to OOM in some scenarios?

Reply:
Or is it handled from the object store side?

Reply:
Yes, it's possible that if the spilling is too slow, the client will still receive an OOM even though enough space will be made eventually.